pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/JavaNode.java

```diff
@@ -42,4 +42,19 @@
     @Deprecated
     void setScope(Scope scope);
 
+    /**
+     * Returns the node representing the type declaration this node is
+     * found in. The type of that node is the type of the {@code this}
+     * expression.
+     *
+     * <p>This returns null for nodes that aren't enclosed in a type declaration.
+     * This includes {@linkplain ASTPackageDeclaration PackageDeclaration},
+     * This includes {@linkplain ASTImportDeclaration ImportDeclaration},
+     * {@linkplain ASTModuleDeclaration ModuleDeclaration},
+     * {@linkplain ASTCompilationUnit CompilationUnit}, and top-level
+     * {@linkplain ASTAnyTypeDeclaration AnyTypeDeclaration}s.
+     *
+     * @return the node of the enclosing type or <code>null</code>.
+     */
+    ASTAnyTypeDeclaration getEnclosingType();
```

Comment:
Also, I'm not sure how sensible it is to add this method to master right now. This cannot consider anonymous class declarations until 7.0.0, so technically it's already violating its contract ("The type of that node is the type of the {@code this} expression").

Reply:
Hm... we could mention this in the javadoc comment, that this doesn't work yet with anonymous classes with PMD 6.

But I still think, we need to deprecate in any case `AbstractJavaRule#getDeclaringType`, since this will be gone. However, I don't see any usage in our own rules, so we could just deprecate it without a replacement?

Reply:
> But I still think, we need to deprecate in any case AbstractJavaRule#getDeclaringType

Definitely. But I think adding the replacement is not necessary yet.. It will alsl cause some conflicts when merging back onto java-grammar, if we modify the javadoc

Reply:
Done: 752e5f8ad556d8b8aeb49338f0defb929d9683ae